storage/kv.go

```diff
@@ -0,0 +1,116 @@
+package storage
+
+import (
+	"encoding/binary"
+	"log"
+	"time"
+
+	"github.com/coreos/etcd/storage/backend"
+	"github.com/coreos/etcd/storage/storagepb"
+)
+
+var (
+	batchLimit    = 10000
+	batchInterval = 100 * time.Millisecond
+	keyBucketName = []byte("key")
+)
+
+type store struct {
+	b       backend.Backend
+	kvindex index
+
+	now        uint64 // current index of the store
+	marshalBuf []byte // buffer for marshal protobuf
+}
+
+func newStore(path string) *store {
+	s := &store{
+		b:          backend.New(path, batchInterval, batchLimit),
+		kvindex:    newTreeIndex(),
+		now:        0,
+		marshalBuf: make([]byte, 1024*1024),
+	}
+
+	tx := s.b.BatchTx()
+	tx.Lock()
+	tx.UnsafeCreateBucket(keyBucketName)
+	tx.Unlock()
+	s.b.ForceCommit()
+
+	return s
+}
+
+func (s *store) Put(key, value []byte) {
+	now := s.now + 1
+
+	s.kvindex.Put(key, now)
```

Comment:
should we wait to do this until after the tx.UnsafePut below? ie, after we've gotten through some error cases, such that the two stores don't get out of sync, potentially